common/utils/data.utils.ts

```diff
@@ -0,0 +1,155 @@
+};
+
+export const getTopics = async (): Promise<Topic[]> => {
+  return topics;
+};
+
+const parseData = (data: ArrayBuffer): void => {
+  languages.length = 0;
+  topics.length = 0;
+
+  const workbook = xlsx.read(data, { type: "array" });
+  const sheetName = workbook.SheetNames[0];
+  const worksheet = workbook.Sheets[sheetName];
+  const json = xlsx.utils.sheet_to_json<InputWord>(worksheet, {
+    blankrows: true,
+    defval: "",
+  });
+
+  // find languages
+  Object.keys(json[0]).forEach((language) => {
+    if (!NON_LANGUAGE_FIELDS.includes(language)) {
+      const [languageName, languageCode, rtl] = language.split("_");
+      languages.push({
+        label: languageName,
+        code: languageCode,
+        rtl: rtl !== undefined,
+      });
+    }
+  });
+
+  const topicMap = new Map<string, Topic>();
+  
+  const setTopic = (topic: Topic, map: Map<LanguageCode, Topic>) => {
+    map.set(topic.label, topic);
+    languages.forEach((language) => {
+      topic.words.set(language.code, []);
+    });
+  };
+
+  // find main topics
+  json.forEach((element: InputWord) => {
+    const isMainTopic =
+      element.Title.includes("T") &&
+      element.Bokmål_nb.toLocaleLowerCase() ===
+        element.Tema1.toLocaleLowerCase();
+    if (!isMainTopic) return;
+
+    if (!topicMap.has(element.Tema1)) {
+      const topic: Topic = {
+        id: element.Title,
+        label: element.Tema1,
+        subTopics: new Map(),
+        words: new Map(),
+      };
+      setTopic(topic, topicMap);
+      return;
+    }
+  });
+
+  // find subTopics
+  json.forEach((element: InputWord) => {
+    const isSubTopic =
+      element.Title.includes("T") &&
+      element.Bokmål_nb.toLocaleLowerCase() !==
+        element.Tema1.toLocaleLowerCase();
+    if (!isSubTopic) return;
+    const topic: Topic = {
+      id: element.Title,
+      label: element.Bokmål_nb,
+      subTopics: new Map(),
+      words: new Map(),
+    };
+    setTopic(topic, topicMap.get(element.Tema1)!.subTopics);
+  });
+
+  // fill words into topics
+  json.forEach((element: InputWord) => {
+    if (!element.Title.includes("V")) return;
+    let images: string[] = [];
+    // find images for the word
+    Object.entries(element).forEach(([key, value]) => {
+      if (key.includes("Bilde")) {
+        images.push(value);
+        return;
+      }
+      return;
+    });
+
+    Object.entries(element).forEach(([key, value]) => {
+      if (NON_LANGUAGE_FIELDS.includes(key)) return;
+      const [_, languageCode] = key.split("_");
+      const word: Word = {
+        id: element.Title,
+        label: value,
+        images: images,
+      };
+      // words within mainTopic
+      if (element.Undertema1 !== element.Tema1) {
+        topicMap
+          .get(element.Tema1)
+          ?.subTopics.get(element.Undertema1)
+          ?.words.get(languageCode)
+          ?.push(word);
+        return;
+      }
+      // words within subTopic
```

Comment:
Comments like these, that explain what is happening, rather than why it's happening, can often be replaced by function calls. `getSubTopicWords` makes more sense to read (at least to me), and it's more likely that the function name will be changed if the code inside changes, than for the comment text to change.